### DISCUSSION TITLE
How to run textual apps in docker?

### DISCUSSION THREAD

**Discussion by mgaitan** [Textualize/textual#2093] [Q&A]
I'm developing my first Textual app . I'm borrowing some code from the "demo" app adapting it to my needs. It's going quite well (and I'm enjoying the experience), but my app must run inside a **docker container**, as part of a complex service orchestrated by `docker-compose` and I've encountered some problems. Here I'm going to show them with the Textual demo app (`python -m textual`) 

The first one, which I managed to get around, is that in order to make the app look really nice, I had to set `COLORTERM: truecolor` as an environment variable to the container, since it originally looked like this: 

![Captura standard](https://user-images.githubusercontent.com/2355719/226447544-80f0d917-b3bb-405b-89f7-4039adf8a881.png)

However, once I got over that one I still had these: 

1) There is no support for mouse events.  Is it possible to enable it? 
2) I see blinking "cursor" artifacts. For example if I close and open the "Sidebar", I see the cursor here. 

![Captura desde 2023-03-20 16-31-08](https://user-images.githubusercontent.com/2355719/226447848-10af88aa-843e-456d-84ea-a08fb3098fa6.png)

Thanks in advance

**Comment by davep** [comment]
I have almost zero experience with Docker so can't directly help with this, but it might be useful if you could run `textual diagnose` in the environment and post the output here. Might give some clues.

**Reply by mgaitan** [reply]
Hi davep. Here it is: 


# Textual Diagnostics

## Versions

| Name    | Value  |
|---------|--------|
| Textual | 0.15.1 |
| Rich    | 13.3.2 |

## Python

| Name           | Value                 |
|----------------|-----------------------|
| Version        | 3.8.12                |
| Implementation | CPython               |
| Compiler       | GCC 10.2.1 20210110   |
| Executable     | /usr/local/bin/python |

## Operating System

| Name    | Value                                                            |
|---------|------------------------------------------------------------------|
| System  | Linux                                                            |
| Release | 5.19.0-35-generic                                                |
| Version | #36~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Fri Feb 17 15:17:25 UTC 2 |

## Terminal

| Name                 | Value     |
|----------------------|-----------|
| Terminal Application | *Unknown* |
| TERM                 | xterm     |
| COLORTERM            | truecolor |
| FORCE_COLOR          | *Not set* |
| NO_COLOR             | *Not set* |

## Rich Console options

| Name           | Value                |
|----------------|----------------------|
| size           | width=119, height=33 |
| legacy_windows | False                |
| min_width      | 1                    |
| max_width      | 119                  |
| is_terminal    | True                 |
| encoding       | utf-8                |
| max_height     | 33                   |
| justify        | None                 |
| overflow       | None                 |
| no_wrap        | False                |
| highlight      | None                 |
| markup         | None                 |
| height         | None                 |

**Comment by TomJGooding** [comment]
Could you try running the command: `export TERM=xterm-256color`? 

I'm wondering if you need to set the `TERM` environment variable in the container (though also zero experience with Docker here I'm afraid!)

**Reply by mgaitan** [reply]
I've just tried that (I've noted was a diff in `textual diagnose` between the container and my host), but there are no changes 😞 . The blinking and the lack of mouse are there.

**Comment by mgaitan** [comment]
I have been debugging this problem with a minimal scenario here 
https://github.com/mgaitan/textual-docker-example and strangely I could not reproduce it. 

With all the base images I used (python3.10, python3.10-alpine, and "fastapi 3.8" -the one I use in my app, based on python3.8-) and in all cases the textual behaves fine, with no flickering and correct mouse events. 

Maybe it is a resource problem when more services are started with compose?

**Reply by TomJGooding** [reply]
This would suggest that some other service is interfering with Textual somehow, but I'm not sure the problem is resource usage.

Presumably there's a way of logging events and resource usage with Docker? Hopefully someone with more experience with Docker can help, but will probably need more info about exactly what other services you are running.

**Reply by cknoll** [reply]
@mgaitan I would suggest that you continue to try to create a minimal "working" example, where "working" means: "can reproduce the issue". I suggest doing this by incrementally transforming your `fastapi 3.8`- example (without the issue) into your actual project via copy-and-paste. Somewhere on that way the issue must occur. If you have pinpointed the cause, you might find a workaround or at least ask more precise questions.

As the two mentioned issues (no mouse events/blinking cursor) might have different causes, it might be recommendable to analyze the separately.

**Reply by gergog76** [reply]
@mgaitan Have you sorted this out at the end? (sorry for opening this thread again) What was the problem?  I have the same mose event problem, but so far no clue how to solve it. Also have a minimal app, but that is not work either, so later on i will try with yours.

**Reply by TomJGooding** [reply]
@gergog76 Did you try the suggestions in popey's post below?

**Reply by gergog76** [reply]
> @gergog76 Did you try the suggestions in popey's post below?

Yes, i use the -it and -e switch. 
```
docker run \
  --rm \
  -it \
  -e TERM=xterm-256color \
  localhost/my-container-name
```

**Comment by popey** [comment]
Hello! Sorry to necro an old discussion, but I may have the answer!

tl;dr You have to pass `-it` at a minimum, because without them, Docker doesn't allocate a proper terminal interface. Add `-e TERM=xterm-256color` to get appropriate colour support.

I made a Docker container of a textual app I wrote, and figured out how to fix this:

My app ([sbommage](https://github.com/popey/sbommage)) is used to load an SBOM (Software Bill of Materials) and display it in a tree-like format using Textualize.

For example, if I do not specify `-it`, the command below shows the screenshot beneath:

```
docker run -v $(pwd)/example_sboms:/sboms sbommage /sboms/syft_1.31.0_docker.io_alpine_latest_spdx.json
```

Note in the screenshot, it's only taking up a portion of the screen:

<img width="2056" height="1261" alt="image" src="https://github.com/user-attachments/assets/ce20f021-d121-49a8-82fd-1ff26ac00462" />

Adding `-it` gives us the ability to interact, and it occupies the entire terminal.

```
docker run -it -v $(pwd)/example_sboms:/sboms sbommage /sboms/syft_1.31.0_docker.io_alpine_latest_spdx.json
```

<img width="2056" height="1261" alt="image" src="https://github.com/user-attachments/assets/7a125951-d8e8-4775-a126-7f0d69487ee3" />

However, if we also add `-e TERM=xterm-256color`, we get the right colours.

```
docker run -it -e TERM=xterm-256color -v $(pwd)/example_sboms:/sboms sbommage /sboms/syft_1.31.0_docker.io_alpine_latest_spdx.json
```

<img width="2056" height="1261" alt="image" src="https://github.com/user-attachments/assets/403b29e8-5969-4ab2-982b-31ec8855cf9c" />

**Reply by willmcgugan** [reply]
Nice. You might also want to set `COLORTERM=truecolor`
